slate@2023.08.25 +rocm w/ ROCm 5.7.1 fails: hip/device_gescale.hip.cc:103:27: error: use of overloaded operator '/' is ambiguous (with operand types 'HIP_vector_type<float, 2>' and 'HIP_vector_type<float, 2>')

### Steps to reproduce the issue

`slate@2023.08.25 +rocm` build fails using:
* spack@develop (c9a111946e9d9a77055ca319b1366ce9945e701e from `Tue Feb 27 14:35:06 2024 -0800`)
* Ubuntu 22.04, AMD64
* GCC 11.4.0
* ROCm 5.7.1, installed with apt, registered as external with Spack

<details><summary>Concretization</summary>

<pre>
 -   jvwrih4  slate@2023.08.25%gcc@11.4.0~cuda~ipo+mpi+openmp+rocm+shared~sycl amdgpu_target=gfx908 build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[+]  cuuwnk3      ^blaspp@2023.11.05%gcc@11.4.0~cuda~ipo+openmp+rocm+shared~sycl amdgpu_target=gfx908 build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[e]  ssikap2          ^rocblas@5.7.1%gcc@11.4.0+tensile amdgpu_target=auto build_system=cmake build_type=Release generator=make patches=f67becc arch=linux-ubuntu22.04-x86_64_v3
[+]  vuy6u52      ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64_v3
[+]  m6xdogt          ^curl@8.6.0%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64_v3
[+]  tsxoxvv              ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  jdnziex              ^openssl@3.2.1%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64_v3
[+]  orzhzee                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  lvpywf6          ^ncurses@6.4%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  ak6hpj5          ^zlib-ng@2.1.5%gcc@11.4.0+compat+opt build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  gyvgctv      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  wtqmyvb      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]  blad35z      ^hip@5.7.1%gcc@11.4.0~cuda+rocm build_system=cmake build_type=Release generator=make patches=3f783ae,aee7249,b589a02,c2ee21c arch=linux-ubuntu22.04-x86_64_v3
[e]  5qltj6u      ^hsa-rocr-dev@5.7.1%gcc@11.4.0~asan+image+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[+]  lzophpa      ^lapackpp@2023.11.05%gcc@11.4.0~cuda~ipo+rocm+shared~sycl amdgpu_target=gfx908 build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[e]  2eg364q          ^rocsolver@5.7.1%gcc@11.4.0+optimal amdgpu_target=auto build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[e]  637xd5z      ^llvm-amdgpu@5.7.1%gcc@11.4.0~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=53f9500,9a97712,b66529f arch=linux-ubuntu22.04-x86_64_v3
[+]  dbj7wo2      ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64_v3
[+]  6a3ezky          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-x86_64_v3
[+]  7gtdw7z          ^libfabric@1.20.1%gcc@11.4.0~debug~kdreg~uring build_system=autotools fabrics=sockets,tcp,udp arch=linux-ubuntu22.04-x86_64_v3
[+]  m4c77rz          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  sojgcrc              ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  7f2zoxe              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  xd27xmc          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  4bin4z7              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  xbtzxbw              ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  uzzbkdc          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  5qv7ikg          ^yaksa@0.3%gcc@11.4.0~cuda~rocm build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  isfgk7s              ^autoconf@2.72%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  qkv54hd              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  uaca3vq              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64_v3
[+]  jz67dau                  ^diffutils@3.9%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  kcyhkht                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  spizmby              ^python@3.11.7%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu22.04-x86_64_v3
[+]  mrupjjx                  ^expat@2.6.0%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  75sfst2                      ^libbsd@0.11.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  mnxtea4                          ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  zckvcsr                  ^gettext@0.22.4%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  3w7w7ki                      ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-x86_64_v3
[+]  zvyqqaz                          ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-x86_64_v3
[+]  k5ioakp                          ^zstd@1.5.5%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  qa4bvzv                  ^libffi@3.4.4%gcc@11.4.0 build_system=autotools patches=070b1f3 arch=linux-ubuntu22.04-x86_64_v3
[+]  oljurle                  ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-x86_64_v3
[+]  owpxnfh                  ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64_v3
[+]  lt5gstq                  ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  i3erqqz                  ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  e72y242      ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu22.04-x86_64_v3
[+]  3x6v26a          ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64_v3
[+]  vpzzlp3              ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3
[+]  x6xxeso              ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  vyp7lu7              ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
</pre></details>

Error:
```
$> spack -e . install
...
[+] /opt/rocm-5.7.1 (external hip-5.7.1-blad35z7dwi4cpksdcj4legt57e6m4lp)
[+] /opt/rocm-5.7.1/ (external hsa-rocr-dev-5.7.1-5qltj6ut43ijfijyd4srt3pcbyyh7a55)
[+] /opt/rocm-5.7.1/llvm (external llvm-amdgpu-5.7.1-637xd5zeylueanqimhbs6muvpzvr24f5)
[+] /opt/rocm-5.7.1/ (external rocblas-5.7.1-ssikap2aoytqce2dx4tef3jygl3inm4d)
[+] /opt/rocm-5.7.1 (external rocsolver-5.7.1-2eg364qxv7fz2y2hadpqznso5jsb2cm2)
...
==> Installing slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp [24/24]
==> No binary for slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/08/0894d8669ed88358cc7c4cb7b77d8467336613245a7b843f3504e9224632ce0e.tar.gz
==> No patches needed for slate
==> slate: Executing phase: 'cmake'
==> slate: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

14 errors found in build log:
     457    #define defs_-NOTFOUND 1
     458                 ^
     459    In file included from <built-in>:809:
     460    <command line>:1:14: warning: ISO C99 requires whitespace after the macro name [-Wc99-extensions]
     461    #define defs_-NOTFOUND 1
     462                 ^
  >> 463    /tmp/root/spack-stage/spack-stage-slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp/spack-src/src/hip/device_gescale.hip.cc:103:27: error: use of overloaded ope
            rator '/' is ambiguous (with operand types 'HIP_vector_type<float, 2>' and 'HIP_vector_type<float, 2>')
     464        scalar_t2 mul = numer / denom;
     465                        ~~~~~ ^ ~~~~~
     466    /tmp/root/spack-stage/spack-stage-slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp/spack-src/src/hip/device_gescale.hip.cc:136:6: note: in instantiation of fun
            ction template specialization 'slate::device::gescale<HIP_vector_type<float, 2>, HIP_vector_type<float, 2>>' requested here
     467    void gescale(
     468         ^
     469    /opt/rocm-5.7.1/include/hip/amd_detail/amd_hip_vector_types.h:564:65: note: candidate function
...
     578    /opt/rocm-5.7.1/include/hip/amd_detail/amd_hip_complex.h:259:1: note: candidate function
     579    COMPLEX_MUL_OP_OVERLOAD(hipDoubleComplex)
     580    ^
     581    /opt/rocm-5.7.1/include/hip/amd_detail/amd_hip_complex.h:78:40: note: expanded from macro 'COMPLEX_MUL_OP_OVERLOAD'
     582        __HOST_DEVICE__ static inline type operator*(const type& lhs, const type& rhs) {           \
     583                                           ^
  >> 584    /tmp/root/spack-stage/spack-stage-slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp/spack-src/src/hip/device_gescale_row_col.hip.cc:97:43: error: use of overloa
            ded operator '*' is ambiguous (with operand types 'HIP_vector_type<double, 2>' and 'const HIP_vector_type<double, 2>')
     585                rowA[ j*lda ] = rowA[ j*lda ] * C[ j ];
...
     611    COMPLEX_MUL_OP_OVERLOAD(hipDoubleComplex)
     612    ^
     613    /opt/rocm-5.7.1/include/hip/amd_detail/amd_hip_complex.h:78:40: note: expanded from macro 'COMPLEX_MUL_OP_OVERLOAD'
     614        __HOST_DEVICE__ static inline type operator*(const type& lhs, const type& rhs) {           \
     615                                           ^
     616    1 warning and 6 errors generated when compiling for gfx908.
  >> 617    make[2]: *** [CMakeFiles/slate.dir/build.make:103: src/hip/device_gescale_row_col.hip.cc.o] Error 1
     618    1 warning generated when compiling for gfx908.
     619    1 warning generated when compiling for gfx908.
     620    1 warning generated when compiling for gfx908.
     621    1 warning generated when compiling for gfx908.
     622    In file included from <built-in>:801:
     623    <command line>:1:14: warning: ISO C99 requires whitespace after the macro name [-Wc99-extensions]

     ...

     685    #define defs_-NOTFOUND 1
     686                 ^
     687    1 warning generated when compiling for host.
     688    1 warning generated when compiling for host.
     689    1 warning generated when compiling for host.
     690    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp/spack-build-jvwrih4'
  >> 691    make[1]: *** [CMakeFiles/Makefile2:88: CMakeFiles/slate.dir/all] Error 2
     692    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-slate-2023.08.25-jvwrih4nqgaykdy23brjcz4u4b7v5fyp/spack-build-jvwrih4'
  >> 693    make: *** [Makefile:139: all] Error 2
```

### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (c9a111946e9d9a77055ca319b1366ce9945e701e)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/14426634/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/14426635/spack-build-out.txt)

@G-Ragghianti @mgates3

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/slate/omp.patch
+++ b/var/spack/repos/builtin/packages/slate/omp.patch
@@ -1,0 +1,20 @@
+diff --git a/src/omptarget/device_util.hh b/src/omptarget/device_util.hh
+index e7aa15c3..3968878d 100644
+--- a/src/omptarget/device_util.hh
++++ b/src/omptarget/device_util.hh
+@@ -6,6 +6,8 @@
+ #ifndef SLATE_OMPTARGET_UTIL_HH
+ #define SLATE_OMPTARGET_UTIL_HH
+ 
++#ifdef SLATE_HAVE_OMPTARGET
++
+ #include <math.h>
+ 
+ namespace slate {
+@@ -157,4 +159,6 @@ inline double abs_val(std::complex<double> x)
+ } // namespace device
+ } // namespace slate
+ 
++#endif // SLATE_HAVE_OMPTARGET
++
+ #endif // SLATE_OMPTARGET_UTIL_HH

--- a/var/spack/repos/builtin/packages/slate/package.py
+++ b/var/spack/repos/builtin/packages/slate/package.py
@@ -51,6 +51,8 @@ class Slate(CMakePackage, CudaPackage, ROCmPackage):
         "2020.10.00", sha256="ff58840cdbae2991d100dfbaf3ef2f133fc2f43fc05f207dc5e38a41137882ab"
     )
 
+    patch("omp.patch", when="@2023.11.05")
+
     variant(
         "mpi", default=True, description="Build with MPI support (without MPI is experimental)."
     )

--- a/var/spack/repos/builtin/packages/slate/package.py
+++ b/var/spack/repos/builtin/packages/slate/package.py
@@ -105,6 +105,7 @@ class Slate(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+rocm", when="+sycl", msg=backend_msg)
     conflicts("+cuda", when="+sycl", msg=backend_msg)
     conflicts("+sycl", when="@:2022.07.00", msg="SYCL support requires SLATE version 2023.08.25")
+    conflicts("^hip@5.6.0:", when="@:2023.08.25", msg="Incompatible version of HIP/ROCm")
 
     def cmake_args(self):
         spec = self.spec

--- a/var/spack/repos/builtin/packages/slate/package.py
+++ b/var/spack/repos/builtin/packages/slate/package.py
@@ -27,6 +27,9 @@ class Slate(CMakePackage, CudaPackage, ROCmPackage):
 
     version("master", branch="master")
     version(
+        "2023.11.05", sha256="d3d925adec137ef4b7d876b2d7d0f8f2ff9d8485fa4125454a42f5da4ac026f3"
+    )
+    version(
         "2023.08.25", sha256="0894d8669ed88358cc7c4cb7b77d8467336613245a7b843f3504e9224632ce0e"
     )
     version(
@@ -77,6 +80,7 @@ class Slate(CMakePackage, CudaPackage, ROCmPackage):
     for val in ROCmPackage.amdgpu_targets:
         depends_on("blaspp +rocm amdgpu_target=%s" % val, when="amdgpu_target=%s" % val)
         depends_on("lapackpp +rocm amdgpu_target=%s" % val, when="amdgpu_target=%s" % val)
+    depends_on("lapackpp@2023.11.05:", when="@2023.11.05:")
     depends_on("lapackpp@2023.08.25:", when="@2023.08.25:")
     depends_on("lapackpp@2022.07.00:", when="@2022.07.00:")
     depends_on("lapackpp@2022.05.00:", when="@2022.05.00:")

